### PR TITLE
Add CPU and GPU implementations of cunumeric.solve

### PR DESCRIPTION
**Pull request by almogsegal**

### CODE REVIEW

**Comment by rohany** [comment]
@almogsegal do you have an idea of how close this is to merging in? I need this PR for some benchmarking and actually wrote an initial implementation of those myself in a separate branch. I can put that aside if this will go in soon.

**Comment by almogsegal** [comment]
@rohany, I believe it shouldn't take too much time. @magnatelee should review this change soon and we'll move on with it.

**Comment by rohany** [comment]
Small request -- can you add a `out=` parameter to this function (deviating from numpy slightly) to control whether a new allocation is made?

File: cunumeric/linalg/linalg.py [hunk 1]
```diff
@@ -555,3 +576,19 @@ def _cholesky(a: ndarray, no_tril: bool = False) -> ndarray:
     )
     output._thunk.cholesky(input._thunk, no_tril=no_tril)
     return output
+
+
+def _solve(a: ndarray, b: ndarray) -> ndarray:
+    a_input = a
```
Comment (magnatelee) [inline cunumeric/linalg/linalg.py:582]: is there a reason you want to make a copy?

File: cunumeric/linalg/solve.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    task.add_output(b)
+    task.execute()
+
+
+def solve(output: DeferredArray, a: DeferredArray, b: DeferredArray) -> None:
+
+    runtime = output.runtime
+    context = output.context
+
+    if runtime.num_procs > 1 or runtime.num_gpus > 1:
```
Comment (magnatelee) [inline cunumeric/linalg/solve.py:60]: this is bad. a `solve` implementation should do the right thing regardless of the processor count, but this code doesn't do anything if there's more than one processor. you should instead add broadcasting constraints on inputs and outputs to all the tasks in this file to make sure they are never partitioned. then you don't need this if statement.

File: run_solve_single.sh [hunk 3]
```diff
@@ -0,0 +1 @@
+PYTHONPATH=../legate.core.internal/build/lib/python3.8/site-packages/legate.core-22.3-py3.8.egg/ python3 ../legate.core.internal/build/bin/legate --gpus 1 examples/solve.py -m 2048 -n 2048 --nrhs 512
```
Comment (magnatelee) [inline run_solve_single.sh:1]: don't push a script for your local testing

File: src/cunumeric/matrix/solve.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace cunumeric {
+
+class SolveTask : public CuNumericTask<SolveTask> {
+ public:
+  static const int TASK_ID = CUNUMERIC_SOLVE;
+
+ public:
+  static void cpu_variant(legate::TaskContext& context);
+#ifdef LEGATE_USE_CUDA
```
Comment (magnatelee) [inline src/cunumeric/matrix/solve.h:29]: any reason that you don't implement an openmp variant? it should be almost the same as the cpu variant except the `openblas_set_num_threads` call.

File: src/cunumeric/matrix/solve_template.inl [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+struct SolveImplBody;
+
+template <LegateTypeCode CODE>
+struct support_solve : std::false_type {
+};
+template <>
+struct support_solve<LegateTypeCode::DOUBLE_LT> : std::true_type {
+};
+template <>
+struct support_solve<LegateTypeCode::FLOAT_LT> : std::true_type {
```
Comment (magnatelee) [inline src/cunumeric/matrix/solve_template.inl:41]: I think the code doesn't support arrays of the `float16` type even though the python code accepts them. you should either modify the python code to convert them to `float32` and back, or do the conversion in the C++ code.

File: src/cunumeric/matrix/solve.cc [hunk 6]
```diff
@@ -0,0 +1,100 @@
+/* Copyright 2021-2022 NVIDIA Corporation
```
Comment (magnatelee) [inline src/cunumeric/matrix/solve.cc:1]: a nitpick: this file (and all the other new files in this PR) didn't exist in 2021. please update the copyright year accordingly.

File: run_solve_single.sh (hunk 3 above)
Comment (almogsegal) [reply run_solve_single.sh:1]: sorry, that was unintentional.

File: cunumeric/linalg/linalg.py (hunk 1 above)
Comment (almogsegal) [reply cunumeric/linalg/linalg.py:582]: to convert the input arrays to float64.

File: src/cunumeric/matrix/solve_template.inl (hunk 5 above)
Comment (almogsegal) [reply src/cunumeric/matrix/solve_template.inl:41]: https://github.com/nv-legate/cunumeric/pull/518#discussion_r945238447

File: src/cunumeric/matrix/solve.cc (hunk 6 above)
Comment (almogsegal) [reply src/cunumeric/matrix/solve.cc]: changed to 2022 only.

File: src/cunumeric/matrix/solve.h (hunk 4 above)
Comment (almogsegal) [reply src/cunumeric/matrix/solve.h:32]: added OpenMP variant.

File: cunumeric/linalg/solve.py (hunk 2 above)
Comment (almogsegal) [reply cunumeric/linalg/solve.py]: added broadcasting constrains on the input/output arrays and removed this limitation.

**Comment by rohany** [comment]
> Small request -- can you add a out= parameter to this function (deviating from numpy slightly) to control whether a new allocation is made?

Bumping this, it should be a small change.

**Comment by magnatelee** [comment]
All of the following tests crash, whereas NumPy raises an unsupported type exception for the first two and produces valid outputs for the rest:
```
import numpy as np
# every solve call in this code crashes if the following line is uncommented
# import cunumeric as np

a = np.random.rand(5, 5).astype("e")
b = np.random.rand(5).astype("e")
try:
    np.linalg.solve(a, b)
    assert False
except TypeError:
    pass

a = np.random.rand(5, 5).astype("e")
b = np.random.rand(5, 5).astype("e")
try:
    np.linalg.solve(a, b)
    assert False
except TypeError:
    pass

a = np.random.rand(5, 5).astype("f")
b = np.random.rand(5).astype("f")
assert np.linalg.solve(a, b).dtype == np.dtype("f")

a = np.random.rand(5, 5).astype("d")
b = np.random.rand(5, 5).astype("f")
assert np.linalg.solve(a, b).dtype == np.dtype("d")
```

**Comment by magnatelee** [comment]
@almogsegal Please add enough correctness tests for `linalg.solve`

File: src/cunumeric/matrix/solve_omp.cc [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                   n,
+                   nrhs,
+                   reinterpret_cast<__complex__ double*>(a),
+                   reinterpret_cast<__complex__ double*>(b));
+  }
+};
+
+/*static*/ void SolveTask::omp_variant(TaskContext& context)
+{
+  solve_template<VariantKind::OMP>(context);
```
Comment (magnatelee) [inline src/cunumeric/matrix/solve_omp.cc:90]: let's add an `openblas_set_num_threads` call as in this file: https://github.com/nv-legate/cunumeric/blob/branch-22.10/src/cunumeric/matrix/potrf_omp.cc#L75

File: src/cunumeric/matrix/solve_template.inl (hunk 5 above)
Comment (magnatelee) [reply src/cunumeric/matrix/solve_template.inl:41]: I was wrong about `float16` and NumPy doesn't seem to support it. Nevertheless, the current code doesn't handle the type error correctly. FYI, `np.dytpe(np.float16).kind == "f"`

**Comment by magnatelee** [comment]
@almogsegal to reproduce the crashes I reported above, you should run it with `LEGATE_TEST=1` and `-cunumeric:test`.

File: examples/solve.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    parser.add_argument(
+        "-t",
+        "--type",
+        default="float64",
+        choices=["float32", "float64", "complex64", "complex128"],
+        dest="dtype",
+        help="data type",
+    )
+    args, unknown = parser.parse_known_args()
+    solve(args.m, args.n, args.nrhs, args.dtype)
```
Comment (bryevdv) [inline examples/solve.py:70]: Besides this example, shouldn't there also be new tests under `tests/integration` that make direct numerical comparisons to numpy results?

**Comment by magnatelee** [comment]
Superseded by #568.

**Pull request closed without merging**